internal/core/eval_context.go

```diff
@@ -0,0 +1,44 @@
+package core
+
+import (
+	"encoding/json"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/zclconf/go-cty/cty"
+
+	"github.com/hashicorp/waypoint/internal/pkg/ctystructure"
+)
+
+// evalCtxTemplateProto adds template data to the eval context if tpl has a
+// "TemplateData" field. This does nothing for any other type.
+func evalCtxTemplateProto(ctx *hcl.EvalContext, key string, tpl proto.Message) error {
+	// Get our template data field in our proto message. If we don't have one
+	// then we don't bother doing anything.
+	val := msgField(tpl, "TemplateData")
+	if !val.IsValid() {
+		return nil
+	}
+
+	valBytes, ok := val.Interface().([]byte)
+	if !ok {
+		return nil
+	}
+
+	var raw map[string]interface{}
+	if err := json.Unmarshal(valBytes, &raw); err != nil {
+		return err
```

Comment:
We should probably give them a detailed error here about what parsing failed. Folks are going to pass bare strings accidentally here most likely so we should give them at first. Maybe: `errors.Wrapf(err, "Error parsing TemplateData for '%s'", key)`.

Reply:
Though maybe I'm mistaken and this would already have been validate before being put into the protobuf?

Reply:
Yeah, the JSON is guaranteed because we encode it ourselves in the SDK based on the structure. 😄 Still, I can wrap this in the hopefully-impossible case it happens.